lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,580 @@
+                if (inputArg.isValid())
+                  if (auto sinkOut = fRef.getValue().dyn_cast<BlockArgument>())
+                    // Record the combinational path that was discovered between
+                    // the ports of this module.
+                    portPaths[inputArg.val][inputArg.fieldId].push_back(
+                        Node(sinkOut, fRef.getFieldID()));
+
+                return connect.getDest();
+              })
+              .Case<RegOp, RegResetOp, SubaccessOp>(
+                  [&](auto reg) { return (Value) nullptr; })
+              .Default([&](Operation *op) {
+                if (op->getNumResults() == 1) {
+                  return (Value)op->getResult(0);
+                }
+                return (Value) nullptr;
+              });
+      if (childVal != nullptr) {
+        children.push_back({childVal, owner});
+      }
+    }
+  }
+
+  /// Record all the leaf ground type values for an aggregate type value. The
+  /// Optional `fieldIdsToAdd` can be used to only add field ids that exist in
+  /// the set and ignore all other fields. This is used for MemOp ports.
+  void gatherAggregateLeafs(Value val,
+                            llvm::SmallDenseSet<size_t> fieldIdsToAdd = {}) {
+    auto valType = val.getType().dyn_cast<FIRRTLBaseType>();
+    // Ignore RefType and GroundType Values.
+    if (!valType || valType.isGround())
+      return;
+    SmallVector<Node> worklist;
+    worklist.push_back(Node(val, 0));
+    // Assign a unique storage location for the `val`, which is the aggregate
+    // base Value.
+    auto baseStorageLocId = storageLocId;
+    valToStorageLocIdMap[val] = {baseStorageLocId};
+    storageLocIdToValuesMap[baseStorageLocId] = {val};
+    // Reserve unique storage location ids for all the subfields of `val`.
+    storageLocId += valType.getMaxFieldID() + 1;
+    auto handleSub = [&](Value res, size_t fieldId, bool isGround) {
+      if (isGround) {
+        if (!fieldIdsToAdd.empty())
+          if (!fieldIdsToAdd.contains(fieldId))
+            return;
+        // Compute the storage location id based on the fieldIndex.
+        auto storageLocId = baseStorageLocId + fieldId;
+        // The results of subaccess op can refer to multiple storage locations.
+        valToStorageLocIdMap[res].push_back(storageLocId);
+        storageLocIdToValuesMap[storageLocId].push_back(res);
+        leafToBaseValMap[res] = val;
+      } else
+        worklist.push_back(Node(res, fieldId));
+    };
+    // Traverse the IR, to get all the lowred fields of the aggregate type. This
+    // representation is equivalent to the LowerTypes transformation.
+    while (!worklist.empty()) {
+      Node node = worklist.pop_back_val();
+      for (auto &use : node.val.getUses()) {
+        Operation *op = use.getOwner();
+        TypeSwitch<Operation *>(op)
+            .Case<SubfieldOp>([&](SubfieldOp sub) {
+              auto bundleType = sub.getInput().getType().cast<BundleType>();
```

Comment:
After the `TypedValue` change, I think the `cast` is no longer needed!

```suggestion
              auto bundleType = sub.getInput().getType();
```

Sadly can't do this for the others, although seems like they could maybe be tightened to their respective types since we eagerly `cast<>` them that way anyway!  Since that for now makes it inconsistent, whatever you like :).

Reply:
Removed the cast.